pkg/querier/distributor_queryable.go

```diff
@@ -97,10 +98,22 @@ func (q *distributorQuerier) Select(_ bool, sp *storage.SelectHints, matchers ..
 	// See: https://github.com/prometheus/prometheus/pull/8050
 	if sp != nil && sp.Func == "series" {
 		ms, err := q.distributor.MetricsForLabelMatchers(ctx, model.Time(q.mint), model.Time(q.maxt), matchers...)
+		queryLimiter := limiter.QueryLimiterFromContextWithFallback(ctx)
```

Comment:
I think this change is easier if we do inside the `MetricsForLabelMatchers`:

https://github.com/aws-observability/cortex-dev/blob/fc862f9c387bbc4f615d79d1a7854d0d3d21aa89/pkg/distributor/distributor.go#L973-L976

Another possible memory optimization on this code would be to not wait all `resps` and dedup those resps as soon as it arrive inside the `ForReplicationSet`

Reply:
Thanks for the feedback.